core/state-machine/src/backend.rs

```diff
@@ -283,29 +357,66 @@ impl<H: Hasher> Clone for InMemory<H> {
 impl<H: Hasher> PartialEq for InMemory<H> {
 	fn eq(&self, other: &Self) -> bool {
 		self.inner.eq(&other.inner)
+			&& self.kv().eq(other.kv()) 
 	}
 }
 
 impl<H: Hasher> InMemory<H> {
+	/// Key value inmemory storage is either used from the
+	/// state db or its own field.
+	/// This function access the right field.
+	fn kv(&self) -> &InMemoryKvBackend {
+		if let Some(kv) = self.kv.as_ref() {
+			kv
+		} else {
+			self.state.as_ref().unwrap().kv_backend()
+		}
```

Comment:
this else condition doesn't make sense, state is only used in the function for returning the immutable reference in as_state_backend.

Reply:
I remember why I use it this way, it was simply to avoid copy when producing the 'as_state_backend' : instead of cloning the values as we do for the trie part it is moving the container.
So after using `as_state_backend` this is getting use instead of the kv field. I could probably switch to simply cloning the `kv` field.

Reply:
ok this is correct indeed

Reply:
would be good to have some comment about this complex behavior